File: leet-code/sortList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
function ListNode (val, next) {
  this.val = val === undefined ? 0 : val
  this.next = next === undefined ? null : next
}
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function (head) {
  if (head == null) {
    return head
  }

  const halfCut = function (link) {
    console.log('halfcut in--', link)
    let fast = link
    let slow = link
    while (fast && fast.next && fast.next.next) {
      fast = fast.next.next
      slow = slow.next
    }
    const lr = slow.next
    slow.next = null
    console.log('halfcut--', link, lr)
    return { ll: link, lr }
  }

  const merge = function (l1, l2) {
    console.log('merge in---', l1, l2)
    let prev = new ListNode()
    const head = prev

    while (l1 || l2) {
      if (l1 == null) {
        prev.next = l2
        break
      } else if (l2 == null) {
        prev.next = l1
        break
      }

      if (l1.val < l2.val) {
        prev.next = l1
        l1 = l1.next
      } else {
        prev.next = l2
        l2 = l2.next
      }
      prev = prev.next
    }
    console.log('merge out---', JSON.stringify(head.next))
    return head.next
  }

  const dfs = function (head) {
    // 如果长度是1，直接返回
    if (head.next == null) {
      return head
    }
    // 分割成左右两部分
    const { ll, lr } = halfCut(head)
    return merge(dfs(ll), dfs(lr))
  }
  const res = dfs(head)
  console.log(res)
}

var head = {
  val: 4,
  next: { val: 2, next: { val: 1, next: { val: 3, next: null } } }
}
console.log(sortList(head))

File: leet-code/sortList.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func sortList(head *ListNode) *ListNode {
	if head == nil {
		return head
	}
	var merge = func(l1, l2 *ListNode) *ListNode {
		prev := &ListNode{-1, nil}
		head := prev

		for l1 != nil || l2 != nil {
			if l1 == nil {
				prev.Next = l2
				break
			} else if l2 == nil {
				prev.Next = l1
				break
			}

			if l1.Val < l2.Val {
				prev.Next = l1
				l1 = l1.Next
			} else {
				prev.Next = l2
				l2 = l2.Next
			}
			prev = prev.Next
		}
		return head.Next
	}
	var halfCount = func(link *ListNode) (rl, rr *ListNode) {
		fast, slow := link, link
		for fast != nil && fast.Next != nil && fast.Next.Next != nil {
			fast = fast.Next.Next
			slow = slow.Next
		}
		lr := slow.Next
		slow.Next = nil
		return link, lr
	}
	var dfs func(head *ListNode) *ListNode
	dfs = func(head *ListNode) *ListNode {
		if head.Next == nil {
			return head
		}
		ll, lr := halfCount(head)
		return merge(dfs(ll), dfs(lr))
	}
	return dfs(head)
}

func sortList2(head *ListNode) *ListNode {
	if head == nil {
		return head
	}
	var merge = func(l1, l2 *ListNode) *ListNode {
		prev := &ListNode{-1, nil}
		head := prev

		for l1 != nil || l2 != nil {
			if l1 == nil {
				prev.Next = l2
				break
			} else if l2 == nil {
				prev.Next = l1
				break
			}

			if l1.Val < l2.Val {
				prev.Next = l1
				l1 = l1.Next
			} else {
				prev.Next = l2
				l2 = l2.Next
			}
			prev = prev.Next
		}
		return head.Next
	}
	var dfs func(head *ListNode) *ListNode
	dfs = func(head *ListNode) *ListNode {
		if head.Next == nil {
			return head
		}
		fast, slow := head, head
		for fast != nil && fast.Next != nil && fast.Next.Next != nil {
			fast = fast.Next.Next
			slow = slow.Next
		}
		lr := slow.Next
		slow.Next = nil

		return merge(dfs(head), dfs(lr))
	}
	return dfs(head)
}

func sortList3(head *ListNode) *ListNode {
	if head == nil {
		return head
	}
	var merge = func(l1, l2 *ListNode) *ListNode {
		prev := &ListNode{-1, nil}
		head := prev

		for l1 != nil || l2 != nil {
			if l1.Val < l2.Val {
				prev.Next = l1
				l1 = l1.Next
			} else {
				prev.Next = l2
				l2 = l2.Next
			}
			prev = prev.Next
		}
		if l1 == nil {
			prev.Next = l2
		} else if l2 == nil {
			prev.Next = l1
		}
		return head.Next
	}
	var dfs func(head *ListNode) *ListNode
	dfs = func(head *ListNode) *ListNode {
		if head.Next == nil {
			return head
		}
		fast, slow := head, head
		for fast != nil && fast.Next != nil && fast.Next.Next != nil {
			fast = fast.Next.Next
			slow = slow.Next
		}
		lr := slow.Next
		slow.Next = nil

		return merge(dfs(head), dfs(lr))
	}
	return dfs(head)
}
